### DISCUSSION TITLE
Trame template

### DISCUSSION THREAD

**Discussion by HuyNguyenDePaul** [Kitware/trame#587] [Trame/VTK]
Hi,

I have been trying for the past couple of days to understand exactly what Trame does when it generates its templates for Vue.js to drive. I think I was able to successfully get the actual HTML template that Trame generates for Vue.js, but I am unable to successfully have another server act as the frontend (such as Django) and have a Trame server connect to that frontend server. If possible, could someone tell me if what I am doing is even feasible. I currently think it might be better to just focus on writing the client manually since I was able to do that successfully and it seems a lot easier, and this seems to be a lot more complicated. But if someone else could guide me, that would be great.

Here is the code that I am using to traverse the HTML elements/template. 
```
def capture_template(layout):
    def traverse(element):
        if hasattr(element, 'html'):
            return element.html
        elif isinstance(element, str):
            return element
        elif isinstance(element, list):
            return ''.join(traverse(child) for child in element)
        else:
            return ''

    return traverse(layout.root)
```

**Comment by jourdain** [comment]
You can see the generated template by just printing it (`print(layout)` or `print(layout.root.html)`).

But either way, trame expect a websocket connection to the actual server process. If you aim to deploy the client side on a django/regular-webserver, it will not work unless you fully handle the launching process + ws connection part and serve the static content like we do when we bundle apps with docker. 

Trame is not like a regular web server. It is more like creating desktop apps. Then it is up to you to make them available in the cloud or just run them locally.

**Reply by HuyNguyenDePaul** [reply]
Ah as expected, this is much more complicated than initially thought. Thank you for the response, seems like writing the JavaScript/client manually is the best way to go. 

I also had a question, for advanced training in VTK or Trame, if we were to do a "private team training", would students receive discounts as they do public courses? I mainly ask this since I wanted some advance training with VTK, but it does not seem like a public course is available.

**Reply by jourdain** [reply]
If you go with private training, you pay for the preparation and presentation time independently of the number of attendees. Also you can focus on specific content relevant to you. In term of discount, I'm not sure but feel free to reach out to [Kitware](https://www.kitware.com/contact/) stating what you are after and we can work with you to figure out what is best for you.
